cli/src/commands/git.rs

```diff
@@ -662,18 +662,11 @@ fn cmd_git_push(
     let mut tx = workspace_command.start_transaction();
     let tx_description;
     let mut branch_updates = vec![];
-    if args.all {
+    if args.all || args.tracked || args.deleted {
         for (branch_name, targets) in repo.view().local_remote_branches(&remote) {
-            match classify_branch_update(branch_name, &remote, targets) {
-                Ok(Some(update)) => branch_updates.push((branch_name.to_owned(), update)),
-                Ok(None) => {}
-                Err(reason) => reason.print(ui)?,
-            }
-        }
-        tx_description = format!("push all branches to git remote {remote}");
-    } else if args.tracked {
-        for (branch_name, targets) in repo.view().local_remote_branches(&remote) {
-            if !targets.remote_ref.is_tracking() {
+            if (args.tracked && !targets.remote_ref.is_tracking())
+                || (args.deleted && targets.local_target.is_present())
+            {
```

Comment:
This goes against 7af45819b. Well, I find it's uneasy to follow.

If we want to deduplicate them, maybe we can extract the code inside the loop to a closure. Then, the loop can be rewritten as `.filter(..).try_for_each(closure)?;`.

Reply:
I have a version like that too, I think, if I haven't abandoned the commit. I thought that repeating it three times is too many, but I don't feel strongly about this; I'll either change it or just get rid of this commit.

Reply:
I'd rather avoid repeating `if args.tracked`/`args.deleted`, etc.`

Reply:
Again, I don't feel strongly here, but I'll explain why I prefer the other option.

I find the question of "what's the same and what's different between these codepaths" important. With your version, it took me a bit to mentally diff the three codepaths and conclude that they are (mostly) the same. I'm simply trying to make it less mental work to get to this point.

So, I like the version where it's obvious what the commonalities are and what the differences are between the options.

In this case, I find you preference a little hard to understand, and therefore curious. It might mean you interpret code in quite a different way than I do, and some other question is more important to you.

I think the version with a closure is a fine compromise, and (again) I don't really mind leaving it as it was.

Reply:
In my mindset, `args.all`/`.tracked`/`.deleted` are UI variables, which tend to change following UI requirements, so I prefer mapping them to more implementation-oriented variables as early as possible.

```rust
let something = if args.all {
} else if args.tracked {
} else if args.deleted {
};
do_stuff_with(something);
```

Perhaps, `something` can be `(boxed_iter_of_filtered_branches, tx_description)` pair, but it might look a little bit uglier if dyn object is involved. So I think deduplicating the loop body would be better.

Another thing I tend to avoid is string construction by parameterizing verb/adjective. Even though we don't have i18n support, it makes harder to grep the codebase.

Reply:
I removed the commit from this PR for now, might submit a version of it later. For reference, it's available at https://github.com/ilyagr/jj/tree/push-tracked-misc for not.